docs/adr/005-publishing-api-debounce.md

```diff
@@ -0,0 +1,121 @@
+  unacceptable
+- The amount of errors is polluting our error management tool (and muting/ignoring them may mask
+  future unrelated bugs)
+
+### Reduce worker count to 1 to enfore in-order processing
+
+In theory, limiting incoming content update processing to a single worker in Search API v2 could
+bypass this problem, as under those circumstances RabbitMQ will guarantee in-order delivery and the
+Discovery Engine API calls would have enough breathing room to not run into rate limiting.
+
+However, this isn't acceptable as:
+- we would lose redundancy in case of worker/Kubernetes node failure
+- a single worker would struggle to keep up with peak load (or full re-sync scenarios) leading to
+  exceeding our desired timeframe of how long it takes for an updated document to be visible in
+  search (<1 minute)
+
+### Implement a locking mechanism
+
+A relatively basic locking mechanism using the existing shared Publishing AWS ElastiCache Redis
+instance (to be migrated to a separate instance in the long term once Platforms have a plan for
+these) as part of the environment would allow us to reach an acceptable level of correctness without
+excessive architectural efforts.
+
+For each document, we would keep up to two keys in Redis:
+- `lock:<content_id>`: Whether a worker is currently processing the document (key only present if
+  locked)
+- `latest:<content_id>`: The last successfully processed version for this document within the last
+  24 hours, if any (exact expiry TBC)
+  - Note: We cannot just rely on the remote document version in Discovery Engine as the document may
+    have been deleted as part of a previous operation
+
+Assuming a peak document update load of 100,000 documents in a 24 hour period, and up to 200 bytes
+for each set of Redis keys/values including allowance for overhead, we would store a maximum of
+20MiB in Redis, meaning a minimally provisioned instance would be sufficient.
+
+If the locking process fails due to Redis being unavailable, we can temporarily fall back onto the
+existing "first update wins" behaviour and track the resulting errors, and perform a full resync if
+the scale of the downtime is large enough that a significant number of updates failed due to the old
+behaviour.
+
+Upon receiving a message, a worker would:
+- use the [`redlock-rb` gem][redlock-gem] to acquire a lock of `lock:<content_id>` with a sensible
+  timeout
+- if the lock currently exists, [wait for it to be available again](#addendum-how-to-wait-for-lock)
+- check `latest:<content_id>` to see if a newer message has recently been successfully processed
+  - if so, do nothing and go straight to releasing the lock
+- process the message (update or delete the document in Discovery Engine)
+  - if processing fails, handle as usual (requeue or discard)
+  - otherwise use `SET latest:<content_id> <version>` to update the latest processed version
+- release the lock
```

Comment:
Is there still a scenario here where we could exceed Discovery Engine's rate
limit if the messages come in quickly in the correct order? Say:

| Time(ms) | Payload Version | Content ID |
|----------|-----------------|------------|
|    0     |               0 | cafebabe   |
|    10    |               1 | cafebabe   |
|    20    |               2 | cafebabe   |
|    30    |               3 | cafebabe   |
|    40    |               4 | cafebabe   |
|    50    |               5 | cafebabe   |

If each of thos messages gets processed quickly enough, we'll hit the rate limit.

Do we want to put some kind of deliberate delay in to handle that case?

For example, if we made the locks timeout after 1 second, and then just didn't
release them explicitly, that would throttle messages to one message per
second.

If we do release the locks explicitly, then this won't necessarily slow down
the processing of sequential messages by much.